File: backend/internal/service/orders_08.go
package service


import (
	"context"
	"time"

	"github.com/biruk-ak/clinical-lis/backend/internal/models"
	"github.com/biruk-ak/clinical-lis/backend/internal/repository"
	"github.com/google/uuid"
)

// OrderRecord08Service encapsulates business rules for OrderRecord08.
type OrderRecord08Service struct {
	repo *repository.OrderRecord08Repository
}

// NewOrderRecord08Service wires the service.
func NewOrderRecord08Service(repo *repository.OrderRecord08Repository) *OrderRecord08Service {
	return &OrderRecord08Service{repo: repo}
}

// Create validates and persists a new OrderRecord08.
func (s *OrderRecord08Service) Create(ctx context.Context, actor, facilityID, code, display, status string, priority int, notes string) (*models.OrderRecord08, error) {
	now := time.Now().UTC()
	rec := &models.OrderRecord08{
		ID:           uuid.NewString(),
		ExternalCode: code,
		DisplayName:  display,
		Status:       status,
		Priority:     priority,
		FacilityID:   facilityID,
		CreatedBy:    actor,
		UpdatedBy:    actor,
		Notes:        notes,
		MetadataJSON: "{}",
		Version:      1,
		IsActive:     true,
		CreatedAt:    now,
		UpdatedAt:    now,
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	if err := s.repo.Create(ctx, rec); err != nil {
		return nil, err
	}
	return rec, nil
}

// Get returns a OrderRecord08 by id.
func (s *OrderRecord08Service) Get(ctx context.Context, id string) (*models.OrderRecord08, error) {
	return s.repo.GetByID(ctx, id)
}

// List returns facility-scoped OrderRecord08 pages.
func (s *OrderRecord08Service) List(ctx context.Context, facilityID string, limit, offset int) ([]models.OrderRecord08, error) {
	if limit <= 0 || limit > 200 {
		limit = 50
	}
	if offset < 0 {
		offset = 0
	}
	return s.repo.List(ctx, facilityID, limit, offset)
}

// Update mutates an existing OrderRecord08.
func (s *OrderRecord08Service) Update(ctx context.Context, actor string, rec *models.OrderRecord08) error {
	if err := rec.Validate(); err != nil {
		return err
	}
	rec.Touch(actor, time.Now().UTC())
	return s.repo.Update(ctx, rec)
}

// Archive soft-deletes a OrderRecord08.
func (s *OrderRecord08Service) Archive(ctx context.Context, id, actor string) error {
	return s.repo.Delete(ctx, id, actor)
}

// Search finds OrderRecord08 by external code prefix.
func (s *OrderRecord08Service) Search(ctx context.Context, facilityID, prefix string) ([]models.OrderRecord08, error) {
	return s.repo.SearchByCode(ctx, facilityID, prefix, 100)
}

// Stats returns active counts for dashboards.
func (s *OrderRecord08Service) Stats(ctx context.Context, facilityID string) (int64, error) {
	return s.repo.CountActive(ctx, facilityID)
}

File: frontend/src/types/orders08.ts
/** Domain types for orders / OrderRecord08 */
export type OrderRecord08Status = 'draft' | 'active' | 'pending' | 'completed' | 'cancelled' | 'archived';

export interface OrderRecord08 {
  id: string;
  externalCode: string;
  displayName: string;
  status: OrderRecord08Status;
  priority: number;
  facilityId: string;
  createdBy: string;
  updatedBy: string;
  notes: string;
  metadataJson: string;
  version: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  archivedAt?: string | null;
}

export interface OrderRecord08CreateRequest {
  externalCode: string;
  displayName: string;
  status: OrderRecord08Status;
  priority: number;
  facilityId: string;
  notes?: string;
}

export interface OrderRecord08ListResponse {
  items: OrderRecord08[];
  total?: number;
}

export interface OrderRecord08StatsResponse {
  activeCount: number;
}

export function isOrderRecord08Active(rec: OrderRecord08): boolean {
  return rec.isActive && rec.status !== 'archived' && rec.status !== 'cancelled';
}

export function formatOrderRecord08Label(rec: OrderRecord08): string {
  return `${rec.externalCode} — ${rec.displayName}`;
}

export function compareOrderRecord08ByPriority(a: OrderRecord08, b: OrderRecord08): number {
  if (a.priority !== b.priority) return b.priority - a.priority;
  return a.displayName.localeCompare(b.displayName);
}
